core/src/main/java/org/elasticsearch/index/translog/Translog.java

```diff
@@ -1348,6 +1348,28 @@ public static void writeOperationNoSize(BufferedChecksumStreamOutput out, Transl
     }
 
     /**
+     * Gets the minimum generation that could contain the sequence number, or the current generation if there is no generation with the
+     * specified sequence number between the minimum and maximum sequence numbers.
+     *
+     * @param seqNo the sequence number
+     * @return the minimum generation for the sequence number, or the current generation
+     */
+    public TranslogGeneration getMinGenerationForSeqNo(final long seqNo) {
+        try (ReleasableLock ignored = writeLock.acquire()) {
+            final long minTranslogFileGeneration = readers
+                    .stream()
+                    .filter(r -> {
+                        final Checkpoint checkpoint = r.getCheckpoint();
+                        return checkpoint.minSeqNo <= seqNo && seqNo <= checkpoint.maxSeqNo;
+                    })
+                    .mapToLong(TranslogReader::getGeneration)
+                    .min()
+                    .orElseGet(this::currentFileGeneration);
```

Comment:
I think another point is that when we are flushing the engine, we ask the translog for the minimum generation for the local checkpoint *plus one* which might not yet be in the translog.